### PR TITLE
Bump hyper

### LINKED ISSUE #1774: update to latest hyper
Need to update to latest hyper, which now includes @meowjesty fix. It introduces other bug fixes which we might encounter.

### PR DESCRIPTION
**Pull request by Razz4780**
Closes #1774 

Notes:
* `hyper` updated to `1.0.0-rc.4`
* Added `tokio-compat` crate with wrappers over `tokio` types, since new `hyper` uses custom IO and executor traits. It's essentially a copy of the unreleased [`hyper-util`](https://github.com/hyperium/hyper-util)
* Had to change `HyperHandler` and `HttpV1` structs a bit, since `hyper::service::Service::call` now takes `&self` instead of `&mut self`. the  @meowjesty could you take a look if everything's fine? The docs say:
```
    /// call takes a &self instead of a mut &self because:
    /// - It prepares the way for async fn,
    ///   since then the future only borrows &self, and thus a Service can concurrently handle
    ///   multiple outstanding requests at once.
    /// - It's clearer that Services can likely be cloned
    /// - To share state across clones you generally need Arc<Mutex<_>>
    ///   that means you're not really using the &mut self and could do with a &self
    /// To see the discussion on this see: https://github.com/hyperium/hyper/issues/3040
```

### CODE REVIEW

**Comment by aviramha** [comment]
Can't we install the tokio compat via git?

**Comment by Razz4780** [comment]
> Can't we add the tokio compat dependency via git?

Sure, we can

**Comment by aviramha** [comment]
> > Can't we add the tokio compat dependency via git?
> 
> Sure, we can

This is preferred and makes more sense imo, with a comment to change it to a release when it is released.

**Comment by Razz4780** [comment]
> > > Can't we add the tokio compat dependency via git?
> > 
> > 
> > Sure, we can
> 
> This is preferred and makes more sense imo, with a comment to change it to a release when it is released.

Done :+1:

File: mirrord/layer/src/tcp_steal/http/v2.rs [hunk 1]
```diff
@@ -14,26 +14,30 @@ use hyper::{
...
-impl<Fut> Executor<Fut> for TokioExecutor
+impl<Fut> Executor<Fut> for DetourGuardExecutor
 where
     Fut: Future + Send + 'static,
     Fut::Output: Send + 'static,
 {
     fn execute(&self, fut: Fut) {
         trace!("starting tokio executor for hyper HTTP/2");
-        tokio::spawn(async move {
-            let _ = DetourGuard::new();
```
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/layer/src/tcp_steal/http/v2.rs:36]: Think I've had something like this before, where it had to be bound to a name, `_` would immediately drop the guard.

**Review by meowjesty — APPROVED** [review]
:bear: :+1: 

Everything looks fine to me.

**Pull request merged**